TSPL.docc/LanguageGuide/Distributed.md

```diff
@@ -0,0 +1,671 @@
+  distributed func greet(name: String) -> String {
+    "Hello \(name)!"
+  }
+}
+```
+
+While the `ActorSystem` remains generic, we did have to specify
+what [`SerializationRequirement`](https://developer.apple.com/documentation/distributed/distributedactorsystem/serializationrequirement)
+the system should be using.
+This is necessary for the type-checking of distributed functions to remain in-tact, and prevents any attempts of using
+distributed actors with actor systems which cannot serialize all of their method's required parameter and result types.
+
+### Emulating callbacks
+
+While in-process actors are able to send and work with closures, this is not possible with distributed actors, as a
+closure
+cannot be serialized over the network. This is one difference in programming for distribution as opposed to local-only
+concurrency with actors.
+
+Let us consider the following "call me later" example expressed in pure local actors:
+
+```swift
+actor Alice {
+  func call(later: @Sendable (String) -> ()) async -> String {
+    Task.detached {
+      // do some asynchronous processing AFTER returning from the 'call' method...
+      later("Here's the info you asked for!")
+    }
+
+    return "Thanks for your call!"
+  }
+}
+```
+
+```swift
+actor Bob {
+  func test(alice: Alice) async {
+    let immediateReply = await alice.call { laterReply in
+      print("later: \(laterReply)")
+    }
+
+    print("immediately: \(immediateReply)")
+  }
+}
+
+// OUTPUT: 
+// immediately: Thanks for your call!
+// later: Here's the info you asked for!
+```
+
+This example models a situation where we call the `Alice` actor, and receive a reply immediately, but at the same time
+we registered a closure to be called at some later point in time when some additional information was looked up by the
+actor. In local-only actors, a common way of modeling this is passing a closure to the actor.
+
+In distributed actors, the same can be achieved using distributed method calls rather than closures, like this:
+
+```swift
+protocol InfoListener: Codable {
+  func additionalInfo(_ info: String) async throws
+}
+
+distributed actor Alice {
+  distributed func call(later: InfoListener) async -> String {
+    Task.detached {
+      // do some asynchronous processing AFTER returning from the 'call' method...
+      later("Here's the info you asked for!")
+    }
+
+    return "Thanks for your call!"
+  }
+}
+```
+
+So instead of a closure, the `Alice` distributed actor is now accepting an `InfoListener`,
+which we'll implement using a distributed actor on the calling side by having Bob be a distributed actor and conform to
+this protocol:
+
+```swift
+distributed actor Bob {
+  func test(alice: Alice) async {
+    let immediateReply = try await alice.call(self)
+
+    print("immediately: \(immediateReply)")
+  }
+}
+
+extension Bob: InfoListener {
+  distributed func additionalInfo(_ info: String) {
+    print("later: \(laterReply)")
+  }
+}
+
```

Comment:
Using resolvable protocol we could have something like that:

```swift
@Resolvable
protocol InfoListener: DistributedActor, Codable {
    distributed func additionalInfo(_ info: String)
}

distributed actor Alice {
    distributed func call(later: $InfoListener) async -> String {
        Task.detached {
            // do some asynchronous processing AFTER returning from the 'call' method...
            try await later.additionalInfo("Here's the info you asked for!")
        }
        
        return "Thanks for your call!"
    }
}

distributed actor Bob: DefaultDistributedActorSystem.SerializationRequirement {
    func test(alice: Alice) async throws {
        let infoListener = try $InfoListener.resolve(id: self.id, using: self.actorSystem)
        let immediateReply = try await alice.call(later: infoListener)
        
        print("immediately: \(immediateReply)")
    }
}

extension Bob: InfoListener {
    distributed func additionalInfo(_ info: String) {
        print("later: \(info)")
    }
}
```

Reply:
That one's tricky, I would not want $ types in distributed method signatures but perhaps there's no workaround yet... I'll dig into this one